cylc/flow/task_job_mgr.py

```diff
@@ -252,7 +261,41 @@ def submit_task_jobs(self, workflow, itasks, curve_auth,
         done_tasks = bad_tasks
 
         for _, itasks in sorted(auth_itasks.items()):
-            platform = itasks[0].platform
+            # Find the first platform where >1 host has not been tried and
+            # found to be unreachable.
+            # If there are no good hosts for a task then the task submit-fails.
+            for itask in itasks:
+                # If there are any hosts left for this platform which we
+                # have not previously failed to contact with a 255 error.
+                if (
+                    set(itask.platform['hosts']) -
+                    self.task_remote_mgr.bad_hosts
+                ):
+                    platform = itask.platform
+                    out_of_hosts = False
+                    break
+                else:
+                    # If there are no hosts left for this platform.
+                    # Set the task state to submit-failed.
+                    itask.waiting_on_job_prep = False
+                    itask.local_job_file_path = None
+                    self._prep_submit_task_job_error(
+                        workflow, itask, '(remote init)', ''
+                    )
+                    self.bad_hosts.difference_update(itask.platform['hosts'])
```

Comment:
So if we run out of hosts on a platform, we put all the (platform's) hosts back in contention right away, rather than wait for the next periodic bad_hosts clear-out?  I guess that makes sense.

Reply:
The periodic clear-out is supposed to be a fail-safe backup that I'm really hoping is never needed. Doing it this way may become inefficient if a platform becomes seriously flaky, but I think that is a problem for site admins: Our intended use case is a comms blip or a login node being taken temporarily offline.